branchout

```diff
@@ -86,6 +86,69 @@ function branchout_os() {
   echo "${OS}"
 }
 
+# Update all Git URLs in 3 levels .git/config files that match the branchout Git base URL by prefix
+function branchout_relocate() {
+  if [ $# -ne 1 ]; then
+    fail "branchout relocate requires exactly one argument: The new Git URL base to apply to all remotes in all repos that currently use ${BRANCHOUT_GIT_BASEURL}"
+  fi
+
+  NEW_GIT_BASEURL="${1}"
+
+  if [ "${BRANCHOUT_GIT_BASEURL}" == "${NEW_GIT_BASEURL}" ]; then
+    fail "Old prefix and new prefix are the same, no point in continuing: ${BRANCHOUT_GIT_BASEURL}"
+  fi
+
+  echo "Relocating all Git repos from ${BRANCHOUT_GIT_BASEURL} to ${NEW_GIT_BASEURL}"
+  echo "in ${PROJECTION_DIRECTORY} and all nested repos 1 and 2 levels deep."
+
+  # Start with the entire git URL
+  OUR_FULL_URL="$(git --git-dir "${PROJECTION_DIRECTORY}/.git" remote get-url origin 2>/dev/null)"
+  echo "This is the current Git URL for ${PROJECTION_DIRECTORY}: ${OUR_FULL_URL}"
+
+  # Build our new URL
+  NEW_FULL_URL="${OUR_FULL_URL//${BRANCHOUT_GIT_BASEURL}/${NEW_GIT_BASEURL}}"
+  echo "This is the new Git URL for ${PROJECTION_DIRECTORY}: ${NEW_FULL_URL}"
+
+  echo "Validating the Git repo pointed at by the new URL above works..."
+  if git ls-remote "${NEW_FULL_URL}" > /dev/null; then
+    echo "New URL exists and you have access, proceeding..."
+  else
+    fail "New URL is bad, repo does not exist, or you have no access to it: ${NEW_FULL_URL}"
+  fi
+
+  # Collect together all of the Git config files to process
+  CONFIG_FILES=("${PROJECTION_DIRECTORY}/.git/config")
+
+  GROUP_CONFIG_FILES=()
+  while IFS='' read -r gitDirectory; do test -f "${gitDirectory}/config" && GROUP_CONFIG_FILES+=("${gitDirectory}/config"); done < <(find "${PROJECTION_DIRECTORY}" -mindepth 2 -maxdepth 2 -type d -name '.git')
+  if [ ${#GROUP_CONFIG_FILES[@]} -gt 0 ]; then
+    CONFIG_FILES=("${CONFIG_FILES[@]}" "${GROUP_CONFIG_FILES[@]}")
+    echo "Found ${#GROUP_CONFIG_FILES[@]} group .git/config files to process"
+  else
+    echo "No group .git/config files found to process."
+  fi
+
+  PROJECT_CONFIG_FILES=()
+  while IFS='' read -r gitDirectory; do test -f "${gitDirectory}/config" && PROJECT_CONFIG_FILES+=("${gitDirectory}/config"); done < <(find "${PROJECTION_DIRECTORY}" -mindepth 3 -maxdepth 3 -type d -name '.git')
+  if [ ${#PROJECT_CONFIG_FILES[@]} -gt 0 ]; then
+    CONFIG_FILES=("${CONFIG_FILES[@]}" "${PROJECT_CONFIG_FILES[@]}")
```

Comment:
I get tripped up by this all the time, does the quoting here stop the expansion of the elements i.e. does this work for n project files where n > 2, ditto for the group ones

Reply:
I don't think it does as the correct number end up in the final array - what makes you say more than 2? Why is 2 not enough to prove it correct? Surely if it was 2 (which it is in the test for group) it would end up as 1 and the total would be 3, not 4 as expected?

Reply:
maybe just one more test to be 100% that the list resolve obviously correctly the numbers a too close to understand at a glance

Reply:
As discussed, there is a subtle hole in the existing test that doesn't cover one of the blocks of script. As agreed, I'll add

1 more group
3 more projects under it

And cross my fingers for a PR approval :-D